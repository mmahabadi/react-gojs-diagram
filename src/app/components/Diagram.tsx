import React, { FC, useCallback, useEffect, useRef } from 'react';
import { ReactDiagram } from 'gojs-react';
import { Link, Node } from '../types';
import { DiagramBuilder } from '../services/DiagramBuilder';

type proptypes = {
  nodes: Node[];
  links: Link[];
  selectedNodeKey: string | null;
  onChange: () => void;
};

const Diagram: FC<proptypes> = ({
  nodes,
  links,
  selectedNodeKey,
  onChange,
}) => {
  const diagramRef: React.RefObject<ReactDiagram> = useRef(null);
  console.log('Diagram rendered');

  useEffect(() => {
    const diagram = diagramRef.current?.getDiagram();
    if (!diagram) {
      return;
    }
    if (diagram) {
      diagram.addDiagramListener('SelectionMoved', onChange);
      diagram.addDiagramListener('PartResized', onChange);
      diagram.addDiagramListener('TextEdited', onChange);

      if (selectedNodeKey) {
        const node = diagram.findNodeForKey(selectedNodeKey);
        if (node) {
          diagram.select(node);
        }
      }
    }

    return () => {
      if (diagram) {
        diagram.removeDiagramListener('SelectionMoved', onChange);
        diagram.removeDiagramListener('PartResized', onChange);
        diagram.removeDiagramListener('TextEdited', onChange);
      }
    };
  }, [selectedNodeKey, onChange]);

  const initDiagram = useCallback(() => {
    return new DiagramBuilder()
      .buildNodeTemplate()
      .buildLinkTemplate()
      .generateDiagram();
  }, []);
  
  return (
    <ReactDiagram
      ref={diagramRef}
      initDiagram={initDiagram}
      divClassName="w-full h-full"
      nodeDataArray={nodes}
      linkDataArray={links}
      onModelChange={onChange}
    />
  );
};
export default React.memo(Diagram);
