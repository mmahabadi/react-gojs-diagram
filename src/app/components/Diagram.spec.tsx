import { RootState, initialState } from '@store';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Diagram from './Diagram';

jest.mock('gojs-react', () => ({
  ReactDiagram: jest.fn().mockImplementation(({ initDiagram, ...props }) => {
    return (
      <div {...props} ref={props.ref}>
        Diagram placeholder
      </div>
    );
  }),
}));

jest.mock('@store', () => ({
  useAppDispatch: () => jest.fn(),
  useAppSelector: jest.fn(() => initialState),
  saveChanges: jest.fn(() => ({ type: 'SAVE_CHANGES' })),
}));

const mockStore = configureStore([]);

describe('Diagram', () => {
  let store: RootState & { dispatch: any };

  const renderComponent = () => {
    return render(
      <Provider store={store as any}>
        <Diagram />
      </Provider>
    );
  };
  beforeEach(() => {
    store = mockStore({
      diagram: initialState,
    }) as any;

    store.dispatch = jest.fn();
  });

  it('renders the Diagram component and matches snapshot', () => {
    const { asFragment } = renderComponent();
    expect(asFragment()).toMatchSnapshot();
  });
});
