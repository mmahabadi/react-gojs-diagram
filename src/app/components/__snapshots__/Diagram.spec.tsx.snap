// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Diagram renders the Diagram component and matches snapshot 1`] = `
<DocumentFragment>
  <div
    divclassname="w-full h-full"
  >
    Diagram placeholder
  </div>
</DocumentFragment>
`;
